### PR TITLE
tracing: Consolidate tracing into a new katatrace package

### LINKED ISSUE #3093: Runtime tracing should reuse Trace() in katautils
There are several similar `trace()` functions in virtcontainers and containerd-shim-v2 (as of https://github.com/kata-containers/runtime/pull/2980) that create spans and set tags for the span. We can reuse `Trace()` in pkg/katautils/tracing.go if we modify it to take tag keys and values as arguments.

Forward port issue: https://github.com/kata-containers/kata-containers/issues/1162

### PR DESCRIPTION
**Pull request by bporter816**
Removes custom trace() functions defined across the project and creates
a single Trace() function in a new package called katatrace. Moves span
tag configuration for static tags to global variables in the package.

Fixes #3093 

Signed-off-by: Benjamin Porter <bporter816@gmail.com>

### CODE REVIEW

**Comment by bporter816** [comment]
@cmaf This is ready for review. It should need the `no-backport-needed` and `forward-port-needed` labels.

Something to point out is that this is a best effort at preserving the existing tags and loggers that were being used to do tracing and report errors. There are certainly some differences (some of the old `trace()` functions didn't report nil context issues at all, for example). We've made some structural changes, mainly moving `Trace()` and tracing.go into a new subpackage of katautils called katatrace. This was done to avoid a cyclical dependency between katautils and virtcontainers. katatracing doesn't import any kata packages.

One bigger change we made was to pass the logger from the caller into `Trace()`, to keep the fields specific to the caller. If that level of detail isn't needed, this could be removed.

We didn't make any changes to the `tracing-test.sh` testing script in the `tests` repo, but it still gave the same output as before our changes. Is there anything that needs to be done there?

File: virtcontainers/pkg/cloud-hypervisor/client/go.mod [hunk 1]
```diff
@@ -3,5 +3,7 @@ module github.com/GIT_USER_ID/GIT_REPO_ID
 require (
 	github.com/antihax/optional v1.0.0
 	golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
-	
+
 )
+
```
Comment (egernst) [inline virtcontainers/pkg/cloud-hypervisor/client/go.mod:8]: Why this change?

File: pkg/katautils/katatrace/tracing.go [hunk 2]
```diff
@@ -90,19 +103,32 @@ func StopTracing(ctx context.Context) {
 
 // Trace creates a new tracing span based on the specified name and parent
 // context.
-func Trace(parent context.Context, name string) (opentracing.Span, context.Context) {
+func Trace(parent context.Context, logger *logrus.Entry, name string, tags ...string) (opentracing.Span, context.Context) {
```
Comment (egernst) [inline pkg/katautils/katatrace/tracing.go:106]: May be worth adding a note in the header re: tags. ie, its a slice of strings, but really its full of key-value pairs, right? Let's just make sure the caller is aware that there should be an even number of entries

File: virtcontainers/api.go [hunk 3]
```diff
@@ -20,7 +21,6 @@ import (
 	"github.com/kata-containers/runtime/virtcontainers/store"
 	"github.com/kata-containers/runtime/virtcontainers/types"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
-	opentracing "github.com/opentracing/opentracing-go"
```
Comment (egernst) [inline virtcontainers/api.go:23]: As we look @ potentially migrating to OpenTelemetry in the future, I appreciate being able to abstract this out of the rest of the code base...

**Review by egernst — COMMENTED** [review]
I'm still working my way through it, but have a couple small bits of feedback in the meantime.

Ideally we'd split this to two commits: one for adding the (new) subpkg, one for updating the rest of the codebase to utilize it (though even there, you could split vc/cli, just for easy quick review). Not a major deal, maybe a nit,, but my preference.

@jodh-intel  can you PTAL too?

**Comment by cmaf** [comment]
@bporter816 could you change the commit to match [this format](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format)? Also a nit but if you could change the subsystem from logging to tracing, that would be helpful.

File: pkg/katautils/config.go [hunk 4]
```diff
@@ -1218,7 +1218,7 @@ func LoadConfiguration(configPath string, ignoreLogging, builtIn bool) (resolved
 	}
 
 	config.Trace = tomlConf.Runtime.Tracing
-	tracing = config.Trace
+	Tracing = config.Trace
```
Comment (cmaf) [inline pkg/katautils/config.go:1221]: If we set the `tracingSet` variable in pkg/katautils/katatrace/tracing.go here, we can remove `tracing bool` from the `CreateTracer()` parameters.

File: pkg/katautils/katatrace/tracing.go [hunk 5]
```diff
@@ -3,34 +3,47 @@
...
+
+// SetLogger sets the logger to be used by katatrace, this fuction is
+// implicitly called by katautils.SetLogger
+func SetLogger(logger *logrus.Entry) {
+	kataTraceLogger = logger
 }
 
 // CreateTracer create a tracer
-func CreateTracer(name string) (opentracing.Tracer, error) {
+func CreateTracer(name string, tracing bool) (opentracing.Tracer, error) {
```
Comment (cmaf) [inline pkg/katautils/katatrace/tracing.go:43]: See above comment; we can remove `tracing bool` if tracingSet is assigned in [pkg/katautils/config.go](https://github.com/kata-containers/runtime/blob/master/pkg/katautils/config.go#L1220-L1221)

**Review by cmaf — CHANGES_REQUESTED** [review]
Thanks @bporter816, a few comments about `CreateTracer()` and the `tracingSet` var

File: virtcontainers/pkg/cloud-hypervisor/client/go.mod (hunk 1 above)
Comment (bporter816) [reply virtcontainers/pkg/cloud-hypervisor/client/go.mod:8]: Not sure about that one...I think it was an automatic editor change. I can revert the `go 1.13` line.

File: pkg/katautils/katatrace/tracing.go (hunk 2 above)
Comment (bporter816) [reply pkg/katautils/katatrace/tracing.go:105]: Fixed

File: pkg/katautils/config.go (hunk 4 above)
Comment (bporter816) [reply pkg/katautils/config.go]: Fixed

File: pkg/katautils/katatrace/tracing.go (hunk 5 above)
Comment (bporter816) [reply pkg/katautils/katatrace/tracing.go]: Fixed

**Comment by bporter816** [comment]
@egernst @cmaf I addressed all the code-specific changes. If that looks good I'll squash and edit the commit format based on those comments.

**Comment by egernst** [comment]
Looks fine to me now @bporter816

**Review by cmaf — APPROVED** [review]
thanks @bporter816

**Comment by cmaf** [comment]
@bporter816 one request, could you format the commits as described in our [docs](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format) and make sure to add a description of the commit? Also, please split it into two commits as @egernst suggested.

**Comment by bporter816** [comment]
@cmaf @egernst I rebased and split the commits. There's a small conflict - not sure what your preference is for resolving those is so I'll leave it as is.

**Comment by cmaf** [comment]
Hi @bporter816, could you please resolve the conflicts and force push the branch again?

**Comment by bporter816** [comment]
@cmaf Done, I think it should be good now.

**Comment by cmaf** [comment]
@egernst @jodh-intel would either of you be able to review?

@bporter816 do you think you will be able to port this to https://github.com/kata-containers/kata-containers? There was a PR that migrates to opentelemetry ([1352](https://github.com/kata-containers/kata-containers/pull/1352)) and another fixing span ordering ([1389](https://github.com/kata-containers/kata-containers/pull/1389)) so the port may require some rework.

**Comment by fidencio** [comment]
Folks, this PR has been stale and I'm closing it as we're walking towards the final 1.13.0 release of the kata-containers-1.x.
From now on, all the development will happen as part of https://github.com/kata-containers/kata-containers/ and fixes should be provided to that repo.

Thanks a lot for your contribution, we're sorry that for some reason it got stale, and we hope to see you around for new adventures in the kata-containers-2.x land.
